VB generators

chsienki opened the pull request:
Enable generators in the Visual Basic compiler

chsienki commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -111,17 +111,30 @@ public override int GetHashCode()
+        }
+
+        [Obsolete("Use GetGenerators(string language) or GetGeneratorsForAllLanguages()")]
         public override ImmutableArray<ISourceGenerator> GetGenerators()
```
📝@333fred I'd like to run this by the API compat review board.

chsienki replied:
In a perfect world, I'd just remove it. Right now I've obsoleted it, and it returns only CSharp generators, which is essentially what it did before.

chsienki commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -111,17 +111,30 @@ public override int GetHashCode()
-            return _diagnosticAnalyzers.GetExtensionsForAllLanguages();
+            // This API returns duplicates of analyzers that support multiple languages.
+            // We explicitly retain this behaviour to ensure back compat
+            return _diagnosticAnalyzers.GetExtensionsForAllLanguages(includeDuplicates: true);
```
📝I'm not sure if it was ever intentional to return duplicate analyzer references, but I've ensured we keep the previous behavior.

jasonmalinowski replied:
Should we just break this? I admit I can't think of anything that would care, and honestly more that might be broken. :-)

chsienki replied:
Yeah, I have no idea. It *is* a public API though, and it's been around for at least 6 years according to blame.

chsienki commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:
```diff
@@ -11,6 +11,43 @@ namespace Microsoft.CodeAnalysis
     public sealed class GeneratorAttribute : Attribute
     {
-        // https://github.com/dotnet/roslyn/issues/: we don't know if we'll keep this, but for now it lets us re-use the analyzer discovery mechanism
+        /// <summary>
```
❓ This is the same layout as the `DiagnosticAnalyzer` attribute, so that we can re-use the lookup logic. Should we enforce this via some stronger mechanism like inheritance (does that break binary compat?) or are we confident with tests covering it working as expected?

jaredpar replied:
I'm fine with it being separate. I think inheritance may be over doing it a bit here.

chsienki commented:
@dotnet/roslyn-compiler for review (/cc @jasonmalinowski for IDE). 

Hopefully it's fairly clear what's going on, but you may find it easier to review commit-by-commit, as I tried to do the work in small distinct stages to make review easier.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:
```diff
@@ -63,6 +64,15 @@ public virtual string Display
+        /// Gets all the diagnostic generators defined in this assembly reference for the given <paramref name="language"/>.
+        /// </summary>
+        /// <param name="language">Language name.</param>
+        public virtual ImmutableArray<ISourceGenerator> GetGenerators(string language) { return ImmutableArray<ISourceGenerator>.Empty; }
```
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGenerators(string language) => ImmutableArray<ISourceGenerator>.Empty; 
```

Nit

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:
```diff
@@ -63,6 +64,15 @@ public virtual string Display
+        public virtual ImmutableArray<ISourceGenerator> GetGeneratorsForAllLanguages() { return ImmutableArray<ISourceGenerator>.Empty; }
+
+        [Obsolete("Use GetGenerators(string language) or GetGeneratorsForAllLanguages()")]
         public virtual ImmutableArray<ISourceGenerator> GetGenerators() { return ImmutableArray<ISourceGenerator>.Empty; }
```
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGenerators() => ImmutableArray<ISourceGenerator>.Empty;
```

Nit

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:
```diff
@@ -63,6 +64,15 @@ public virtual string Display
         /// <summary>
         /// Gets all the source generators defined in this assembly reference.
         /// </summary>
+        public virtual ImmutableArray<ISourceGenerator> GetGeneratorsForAllLanguages() { return ImmutableArray<ISourceGenerator>.Empty; }
```
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGeneratorsForAllLanguages() => ImmutableArray<ISourceGenerator>.Empty;
```

Nit

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs:
```diff
@@ -15,19 +15,16 @@ internal sealed class AdditionalSourcesCollection
 
-        internal AdditionalSourcesCollection()
+        internal AdditionalSourcesCollection(string fileExtension)
         {
```
When dealing with extensions can never remember if we should or should not include the `.`. So `".cs"` vs. `"cs"`. Consider either enforcing one or the other is passed here or normalizing the input to a specific one.

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs:
```diff
@@ -100,11 +99,11 @@ public bool Contains(string hintName)
+        private string AppendExtensionIfRequired(string hintName)
         {
-            if (!hintName.EndsWith(".cs", _hintNameComparison))
+            if (!hintName.EndsWith(_fileExtension, _hintNameComparison))
```
Based on this it seems like we should assert in the ctor that the provided extension begins with `"."`

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:
```diff
@@ -11,6 +11,43 @@ namespace Microsoft.CodeAnalysis
     {
-        // https://github.com/dotnet/roslyn/issues/: we don't know if we'll keep this, but for now it lets us re-use the analyzer discovery mechanism
+        /// <summary>
+        /// The source languages to which this generator applies.  See <see cref="LanguageNames"/>.
```
```suggestion
        /// The source languages to which this generator applies. See <see cref="LanguageNames"/>.
```

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:
```diff
@@ -11,6 +11,43 @@ namespace Microsoft.CodeAnalysis
+        public string[] Languages { get; }
+
+        /// <summary>
+        /// Attribute constructor used to specify the attached class is a source generator that provides CSharp sources.
```
```suggestion
        /// Attribute constructor used to specify the attached class is a source generator that provides CSharp sources.
```

Crazy aside: is it required to be a `class` or can it say be a `struct`?

chsienki replied:
It *has* to be a class. We explicitly set the allowed location of the attribute to be class (see line 11) so loading via AnalyzerReference it can't ever be a struct. 

As a second defense we check that the generator is also an `ISourceGenerator`, so there's no way you can pass a struct via the generator driver in code either.

jaredpar commented on src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCompiler.vb:
```diff
@@ -297,6 +297,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 embeddedFiles.Add(resolvedPath)
             Next
         End Sub
+        Private Protected Overrides Function RunGenerators(input As Compilation, parseOptions As ParseOptions, generators As ImmutableArray(Of ISourceGenerator), analyzerConfigOptionsProvider As AnalyzerConfigOptionsProvider, additionalTexts As ImmutableArray(Of AdditionalText), diagnostics As DiagnosticBag) As Compilation
```
```suggestion

        Private Protected Overrides Function RunGenerators(input As Compilation, parseOptions As ParseOptions, generators As ImmutableArray(Of ISourceGenerator), analyzerConfigOptionsProvider As AnalyzerConfigOptionsProvider, additionalTexts As ImmutableArray(Of AdditionalText), diagnostics As DiagnosticBag) As Compilation
```

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -254,7 +267,33 @@ private static IEnumerable<string> GetDiagnosticsAnalyzerSupportedLanguages(PEMo
+            return peModule.IsTargetAttribute(customAttrHandle, s_generatorAttributeNamespace, nameof(GeneratorAttribute), ctor: out _);
+        }
+
+        private static IEnumerable<string> GetGeneratorSupportedLanguages(PEModule peModule, CustomAttributeHandle customAttrHandle)
```
Could this be a local function?

chsienki replied:
I've left these for now as they've got the comments explaining them and it seemed weird embedding them in the ctor.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -254,7 +267,33 @@ private static IEnumerable<string> GetDiagnosticsAnalyzerSupportedLanguages(PEMo
+            return ReadLanguagesFromAttribute(ref argsReader);
+        }
 
+        private static bool IsGeneratorAttribute(PEModule peModule, CustomAttributeHandle customAttrHandle)
```
Could this be a local function?

chsienki replied:
I refactored out the extensions logic to make it simpler and not need it.

chsienki commented:
Closing/reopening to fix CI

jasonmalinowski reviewed (changes requested):
IDE stuff looks good except it also needs a change in src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs, where we updated in https://github.com/dotnet/roslyn/pull/48400. There's comments with "we don't support VB" that need to be updated with the right diagnostic IDs.

paul1956 commented:
Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?

chsienki commented on src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:
```diff
@@ -248,25 +248,74 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
             Assert.Equal(AnalyzerLoadFailureEventArgs.FailureErrorCode.UnableToCreateAnalyzer, errors.First().ErrorCode);
         }
 
-        // can't load a framework targeting generator, which these are in desktop
```
Ah, not sure how that got removed. Good catch

jaredpar commented:
@paul1956 

> Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?

This is the same as C# source generators, just for VB. There aren't any VB specific docs / examples at this point. The faeture doesn't exist until this is merged so can't really have samples yet ;)

The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

chsienki commented:
@jaredpar 
> The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

Do we want to update the cookbook as part of this PR, or are we ok leaving it until later?

jaredpar commented:
> Do we want to update the cookbook as part of this PR, or are we ok leaving it until later?

I'm indifferent here. Think we need the cookbook updated before this is RTM but I don't think we need to tie it to this PR.

jaredpar commented on src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:
```diff
@@ -248,6 +248,7 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
             Assert.Equal(AnalyzerLoadFailureEventArgs.FailureErrorCode.UnableToCreateAnalyzer, errors.First().ErrorCode);
         }
 
+        // can't load a framework targeting generator, which these are in desktop	
         [ConditionalFact(typeof(CoreClrOnly))]
```
```suggestion
        [ConditionalFact(typeof(CoreClrOnly), Reason = "Can't load a framework targeting generator, which these are in desktop")]
```

AnthonyDGreen commented:
> @paul1956
> 
> > Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?
> 
> This is the same as C# source generators, just for VB. There aren't any VB specific docs / examples at this point. The faeture doesn't exist until this is merged so can't really have samples yet ;)
> 
> The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

You don't say...

chsienki commented:
@jasonmalinowski I've updated the AnalyzerHelper to include the VB warning, and added the required warning to the VB code itself.

I have a test in the VB layer, but currently can't run it because we're not running VB command line tests on core. That is proving to be an _interesting_ challenge, so I may just open an issue for it and add the test as part of the move later on.

chsienki commented:
Ok I've added the test, even though it will always be skipped for now, and opened an issue to track enabling the command-line tests for VB.

paul1956 commented:
When does this appear in some Preview I can try?

chsienki commented:
@paul1956 
> When does this appear in some Preview I can try?

When the PR merges it'll be available in the Roslyn nightly NuGet packages. We don't have a definitive date or target release of visual studio that it'll be available by, but we'll be sure to call it out when there is one.

paul1956 commented:
@chsienki when it is available in NuGet can you post minimal instructions on where to download and simple usage?

chsienki commented:
@dotnet/roslyn-compiler for a second review please :)

cston commented on src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:
```diff
@@ -248,25 +248,74 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
+            Assert.Equal(10, generators.Length);
 
             var typeNames = generators.Select(g => g.GetType().FullName);
             Assert.Contains("Microsoft.CodeAnalysis.UnitTests.AnalyzerFileReferenceTests+TestGenerator", typeNames);
```
Consider using: `AssertEx.SetEqual(new[] { "TestGenerator", "BaseGenerator", ... }, typeNames);` here and below.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -232,29 +241,44 @@ where supportedLanguages.Any()
         private static IEnumerable<string> GetDiagnosticsAnalyzerSupportedLanguages(PEModule peModule, CustomAttributeHandle customAttrHandle)
         {
             // The DiagnosticAnalyzerAttribute has one constructor, which has a string parameter for the
             // first supported language and an array parameter for addition supported languages.
```
>addition [](start = 67, length = 8)

"additional"?

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -232,29 +241,44 @@ where supportedLanguages.Any()
+        private static IEnumerable<string> GetGeneratorSupportedLanguages(PEModule peModule, CustomAttributeHandle customAttrHandle)
+        {
+            // The GeneratorAttribute has two constructors: one default, and one with a string parameter for the
+            // first supported language and an array parameter for addition supported languages.
```
>addition [](start = 67, length = 8)

"additional"?

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -305,37 +322,41 @@ private static string GetFullyQualifiedTypeName(TypeDefinition typeDef, PEModule
-        private sealed class Extensions<TExtension> where TExtension : class
+        private sealed class Extensions<TExtension, TAttribute>
+            where TExtension : class
+            where TAttribute : Attribute
```
>TAttribute [](start = 18, length = 10)

Is this type parameter used outside of the constructor? If not, consider passing the namespace and name explicitly to the constructor, or create a generic `static Extensions<TExtension> Create<TExtension, TAttribute>(...)` factory method.

chsienki replied:
I refactored it just pass in the type as a param instead.

jasonmalinowski commented on src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:
```diff
@@ -248,25 +248,74 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
-            var generators = reference.GetGenerators();
-            Assert.Equal(5, generators.Length);
+            var generators = reference.GetGeneratorsForAllLanguages();
+            Assert.Equal(10, generators.Length);
```
And if you do follow @cston's advice below, delete this assert; nothing more fun than this accidentally returning more than it should and the failure is "Expected: 10, actual: 11" and you have no idea what happened.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -344,7 +365,13 @@ private static ImmutableArray<TExtension> CreateExtensionsForAllLanguages(Extens
-                    builder.AddRange(analyzers);
+                    foreach (var analyzer in analyzers)
+                    {
+                        if (includeDuplicates || !builder.Contains(t => t.GetType().Equals(analyzer.GetType())))
```
>builder.Contains [](start = 50, length = 16)

It looks like this is O(n^2) where n is the number of extensions. Consider adding all analyzers as before and check for duplicates outside the loop: `return builder.ToImmutable().Distinct(comparer);`

chsienki replied:
The irony that I originally did it this way to avoid the duplicate work of calling distinct afterwards. Good catch.

cston commented on src/Compilers/CSharp/Test/Semantic/SourceGeneration/AdditionalSourcesCollectionTests.cs:
```diff
@@ -43,6 +43,20 @@ public void HintName_ValidValues(string hintName)
+            Assert.True(asc.Contains(hintName));
+
+            var sources = asc.ToImmutableAndFree();
+            Assert.True(sources[0].HintName.EndsWith(".vb"));
```
>sources[0] [](start = 24, length = 10)

Consider using `sources.Single()` here and in test above.

cston commented on src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:
```diff
@@ -10257,6 +10257,62 @@ dotnet_diagnostic.{diagnosticId}.severity = {analyzerConfigSeverity}")
+            Dim directory = Temp.CreateDirectory()
+            Dim generatorPath = Path.Combine(directory.Path, "generator.dll")
+            Dim compilation = VisualBasicCompilation.Create($"generator_{targetFramework}",
+                                                            {VisualBasicSyntaxTree.ParseText(generatorSource)},
```
>generatorSource [](start = 93, length = 15)

It looks like the source is C#.

chsienki replied:
It is, generators themselves can be written in any language. I'll switch it over to VB to make it consistent though.

cston commented on src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:
```diff
@@ -248,25 +248,80 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
+
             Assert.DoesNotContain("Microsoft.CodeAnalysis.UnitTests.TestGeneratorNoAttrib", typeNames);
             Assert.DoesNotContain("Microsoft.CodeAnalysis.UnitTests.Test.NotAGenerator", typeNames);
             Assert.DoesNotContain("Microsoft.CodeAnalysis.UnitTests.NotAGenerator", typeNames);
```
`Assert.DoesNotContain()` is no longer necessary.

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -344,10 +362,29 @@ private static ImmutableArray<TExtension> CreateExtensionsForAllLanguages(Extens
+
+            private class ExtTypeComparer : IEqualityComparer<TExtension>
+            {
+                public static ExtTypeComparer Instance = new ExtTypeComparer();
```
>static [](start = 23, length = 6)

`readonly`

cston commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -344,10 +362,29 @@ private static ImmutableArray<TExtension> CreateExtensionsForAllLanguages(Extens
+            {
+                public static ExtTypeComparer Instance = new ExtTypeComparer();
+
+                public bool Equals(TExtension? x, TExtension? y) => x is null && y is null || x is object && y is object && x.GetType().Equals(y.GetType());
```
>x is null && y is null || x is object && y is object && x.GetType().Equals(y.GetType() [](start = 68, length = 86)

Can use `object.Equals(x?.GetType(), y?.GetType())`.
